Add metric for checking common documentation paths

d33bs opened the pull request:
<!-- _modified from [EmbeddedArtistry](https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/)_
_referenced with modifications from [pycytominer](https://github.com/cytomining/pycytominer/blob/main/.github/PULL_REQUEST_TEMPLATE.md)_ -->

# Description

This PR adds a metric which checks for "common documentation" paths within a given repository. We perform this measurement in alignment with #150. As part of measuring how the Almanack itself meets this standard I created a docs directory with a reference to where to find the documentation for this project. As a minor note to this effect, the `book` content is currently stored within the `src` as it is included with the package builds for reading through the Python package.

I feel we could continue to work on #150 with another change after this which involves looking at common GitHub Pages locations based on the git remotes listing of a repository (which would likely yield an HTTP link we could use to form the default GitHub pages default). We could also check for docs badging within the main readme. Either way, I thought these changes might be better outside this PR, as they're a bit different in nature.

References #150

## What is the nature of your change?

- [ ] Content additions or updates (adds or updates content)
- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (these changes would cause existing functionality to not work as expected).

# Checklist

Please ensure that all boxes are checked before indicating that this pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own contributions.
- [x] I have commented my content, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to related documentation (outside of book content).
- [x] My changes generate no new warnings.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests that prove my additions are effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

falquaddoomi commented on tests/metrics/test_data.py:
```diff
@@ -34,21 +35,19 @@ def test_generate_repo_data(entropy_repository_paths: dict[str, pathlib.Path]) -
+
+        # check that all keys exist in the output from metrics
+        # table to received dict
+        assert all(
+            x == y
+            for x, y in zip(
+                sorted(data.keys()),
+                sorted([metric["name"] for metric in metrics_table["metrics"]]),
+            )
+        )
```
This is a very minor nit, but in `pytest` you get *slightly* better performance and a bit more descriptive output on a test failure when you compare collections directly rather than doing the comparison yourself and then checking if `all()` is true. For example, you could rewrite this code as:

```suggestion
        assert sorted(data.keys()) == sorted([metric["name"] for metric in metrics_table["metrics"]])
```

In this case `pytest` will use its [special equality handling](https://docs.pytest.org/en/stable/how-to/assert.html#making-use-of-context-sensitive-comparisons) to do a pairwise check between each element, stopping early if there's a discrepancy, and will report more information, e.g. the index at which the two differed, whether one side contains excess items, etc. The original version would only report the boolean array provided to `all()`.

d33bs replied:
Thanks this is great! I'll add this in.

falquaddoomi commented on tests/metrics/test_data.py:
```diff
@@ -34,21 +35,19 @@ def test_generate_repo_data(entropy_repository_paths: dict[str, pathlib.Path]) -
-        ]
-        assert all(key in data for key in expected_keys)
+        # open the metrics table
+        with open(METRICS_TABLE, "r") as f:
+            metrics_table = yaml.safe_load(f)
```
Cool to see these getting read from the metadata file! 😄

d33bs replied:
Cheers thanks!

falquaddoomi reviewed (commented):
Checking for documentation seems useful, cheers to that. 😄 AFAICT there weren't any issues with your code, but I made a few comments that I hoped might help simplify it by leveraging existing functionality in the libraries you're using.

d33bs commented on src/almanack/git.py:
```diff
@@ -178,38 +178,64 @@ def detect_encoding(blob_data: bytes) -> str:
-    # Decode and return content as a string
-    return blob_data.decode(detect_encoding(blob_data))
+    # Split the file path into parts (e.g., "docs/readme.txt" -> ["docs", "readme.txt"])
+    path_parts = filepath.split("/")
+
+    for i, part in enumerate(path_parts):
+        # Handle case-insensitivity by converting parts to lowercase if needed
+        if case_insensitive:
+            # note: we noqa below to avoid warning about overwritten loop variable
+            part = part.lower()  # noqa: PLW2901
+
+        try:
+            # If case-insensitive, also convert the entry name to lowercase
+            entry = (
+                tree[part]
+                if not case_insensitive
+                else next(e for e in tree if e.name.lower() == part)
+            )
+        except (KeyError, StopIteration):
+            # If the part is not found, return None early
+            return None
+
+        if entry.type == pygit2.GIT_OBJECT_TREE:
+            # If the entry is a tree (directory), navigate into it
+            tree = repo[entry.id]
+        elif entry.type == pygit2.GIT_OBJECT_BLOB:
+            # If it's a blob and either the last part of the path or in the root, read the blob's data
+            if i == len(path_parts) - 1:
+                blob = repo.get(entry.id)  # Retrieve the blob object
+                blob_data: bytes = blob.data
+                decoded_data = blob_data.decode(detect_encoding(blob_data))
+                return decoded_data
+            else:
+                return None
+        else:
+            # If we encounter a non-matching structure, return None
+            return None
+
+    # If the loop completes without finding a blob, the file wasn't found
+    return None
```
Thanks this makes a lot of sense and seems sensible to me! I made a change and will push to help address this comment.

d33bs commented on src/almanack/metrics/data.py:
```diff
@@ -193,6 +195,45 @@ def default_branch_is_not_master(repo: pygit2.Repository) -> bool:
+
+    # Check each documentation path using the find_and_read_file function
+    for doc_path in common_docs_paths:
+        if find_and_read_file(repo=repo, filepath=doc_path) is not None:
```
That's a great idea, thanks! I can see now that I violated the single responsibility principle fairly clearly. I'll update this based on what you mentioned.

d33bs commented on tests/data/almanack/repo_setup/create_repo.py:
```diff
@@ -159,29 +159,21 @@ def repo_setup(
-
-    Returns:
-        pygit2.Repository: The initialized repository with files.
+    Set up a temporary repository with specified files, supporting nested file paths.
```
Ah good point, I think I modified this during an iteration towards fixing issues. I'll reinstate the original with modifications.

d33bs commented:
Thanks @falquaddoomi for the review! I've addressed your comments with various updates. Please let me know if there's anything else I may do to help meet an approval for this PR.

falquaddoomi reviewed (approved):
Looks good to me!

d33bs commented:
Thanks @falquaddoomi ! Merging this in now.
